webapp/src/components/user_settings/audio_devices_settings_section.tsx

```diff
@@ -0,0 +1,264 @@
+    const [selectedOption, setSelectedOption] = useState<SelectOption|null>(null);
+
+    const label = deviceType === 'inputs' ?
+        formatMessage({defaultMessage: 'Microphone input'}) :
+        formatMessage({defaultMessage: 'Speaker output'});
+
+    const options = devices.map((device) => {
+        return {
+            value: device.deviceId,
+            label: device.label,
+        };
+    });
+
+    const getOption = () => {
+        if (selectedOption) {
+            return selectedOption;
+        }
+
+        const defaultDeviceID = deviceType === 'inputs' ?
+            window.localStorage.getItem('calls_default_audio_input') :
+            window.localStorage.getItem('calls_default_audio_output');
+
+        for (const device of devices) {
+            if (device.deviceId === defaultDeviceID) {
+                return {
+                    label: device.label,
+                    value: device.deviceId,
+                };
+            }
+        }
+
+        return {
+            label: devices[0]?.label ?? '',
+            value: devices[0]?.deviceId ?? '',
+        };
+    };
+
+    useImperativeHandle(ref, () => {
+        return {
+            getOption,
+        };
+    }, [devices, selectedOption]);
+
+    return (
+        <SelectionWrapper>
+            <SelectLabel
+                id={name + 'Label'}
+                htmlFor={name + 'Select'}
+            >
+                {label}
+            </SelectLabel>
+            <StyledReactSelect
+                inputId={name + 'Select'}
+                aria-labelledby={name + 'Label'}
+                className='react-select singleSelect'
+                classNamePrefix='react-select'
+                options={options}
+                clearable={false}
+                isClearable={false}
+                isSearchable={false}
+                components={{IndicatorSeparator: () => null}}
+                value={getOption()}
+                onChange={(opt: SelectOption) => setSelectedOption(opt)}
+            />
+        </SelectionWrapper>
+    );
+});
+
+const StyledReactSelect = styled(ReactSelect)`
+  width: 320px;
+`;
+
+const SelectionWrapper = styled.div`
+  display: flex;
+  align-items: center;
+  justify-content: space-between;
+`;
+
+const SelectLabel = styled.label`
+  font-size: 12px;
+  line-height: 16px;
+  margin: 0;
+`;
+
+export default function AudioDevicesSettingsSection() {
+    const {formatMessage} = useIntl();
+    const [active, setActive] = useState(false);
+    const [devices, setDevices] = useState<MediaDeviceInfo[]>([]);
+
+    const title = formatMessage({defaultMessage: 'Audio devices settings'});
+    const subtitle = formatMessage({defaultMessage: 'Setup audio input and output devices used in calls'});
+    const editLabel = formatMessage({defaultMessage: 'Edit'});
+
+    const audioInputsRef = useRef<AudioDevicesSelectionHandle>(null);
+    const audioOutputsRef = useRef<AudioDevicesSelectionHandle>(null);
```

Comment:
Is this necessary? I wonder if there's an easier way (like passing in a `setState` so that the child can save the current selection)?

Reply:
@cpoile I had to choose between those two approaches and didn't really like either. Pushing state from child to parent was equally (if not more) verbose as it would have duplicated more state or worse needed to pass getters as well. In the end I think the ref approach is okaish but if you have a substantial simplification in mind please let me know.